### ISSUE TITLE
False positive with -Wreturn-stack-address and rvalue references

### ISSUE THREAD

**Issue by cameronwhite** [llvm/llvm-project#25730]
|  |  |
| --- | --- |
| Bugzilla Link | [25356](https://llvm.org/bz25356) |
| Version | trunk |
| OS | Linux |
| CC | @DougGregor |

## Extended Description 
The attached code snippet produces the warning below, which I believe is incorrect. If the typedef of T is changed to just "int", no error is produced.

------------

test.cc:12:12: warning: returning address of local temporary object [-Wreturn-stack-address]
    return x;
           ^
test.cc:11:12: note: binding reference variable 'x' here
    auto &&x = foo();
           ^   ~~~~~

------------

typedef int * T;

T foo()
{
    T x = 0;
    return x;
}

T bar()
{
    auto &&x = foo();
    return x;
}

int main(void)
{
    T x = bar();
    return 0;
}

**Comment by wheatman** [comment]
The first case was fixed between 3.8.1 and 3.9.0
https://godbolt.org/z/fj1TTMPev

code
```
typedef int * T;

T foo()
{
    T x = 0;
    return x;
}

T bar()
{
    auto &&x = foo();
    return x;
}

int main(void)
{
    T x = bar();
    return 0;
}
```

old warning
```
<source>:12:12: warning: returning address of local temporary object [-Wreturn-stack-address]
    return x;
           ^
<source>:11:12: note: binding reference variable 'x' here
    auto &&x = foo();
           ^   ~~~~~
<source>:17:7: warning: unused variable 'x' [-Wunused-variable]
    T x = bar();
      ^
2 warnings generated.
Compiler returned: 0
```

The second case was fixed between 6.0.1 and 7.0.0
https://godbolt.org/z/brjWcaY81

code
```
#include <iostream> 
 
 int x = 5;
 int* foo() 
 {   
     return &x;
 }
 
 int& foo2()
 {   
     auto&& cell = foo();
     return *cell;
 }
 
 int main()
 {
     int& tr = foo2 ();
     std::cout << tr << std::endl;
     std::cout << &tr << std::endl;
 }
```

old warning
```
<source>:12:14: warning: returning reference to local temporary object [-Wreturn-stack-address]
     return *cell;
             ^~~~
<source>:11:13: note: binding reference variable 'cell' here
     auto&& cell = foo();
            ^      ~~~~~
1 warning generated.
Compiler returned: 0
```
